lib/brexit_checker/criteria/filter.rb

```diff
@@ -0,0 +1,32 @@
+class BrexitChecker::Criteria::Filter
+  def call(selected_criteria)
+    loop do
+      old_size = selected_criteria.size
+      selected_criteria &&= possible_criteria(selected_criteria)
+      break if selected_criteria.size == old_size
+    end
+
+    selected_criteria
+  end
+
+private
+
+  def possible_criteria(selected_criteria)
+    BrexitChecker::Question.load_all
+      .select { |question| question.show?(selected_criteria) }
+      .flat_map(&:options)
+      .flat_map { |option| possible_values(option, selected_criteria) }
+      .compact
```

Comment:
Indeed, it's quite complex, mainly because of the (potential) hierarchical structure of the options. I think the smell / headache is coming from trying to emulate the implicit show/hide CSS for the sub-options in the backend.

When I was writing the class I did wonder if we could push the option filtering bit into the model, but we've already put our foot in it there with Option#show?. I also toyed with telling options about their parents, but that made the initialisation weird.

I think we should avoid pushing code into the model unless it's a reusable or natural fit, which is unlikely given the algorithmic purpose. But we should do something, so I've tried a few smaller tweaks that I'm hope will make it just about OK.

Could you take another look?